### ISSUE TITLE
UI::push inside attachment listener of a dialog child makes it fail

### ISSUE THREAD

**Issue by mrgreywater** [vaadin/flow#16635]
### Description

When using ui push inside a grid attach listener inside a dialog (while manual push mode is enabled) makes the whole ui fail to load.

### Outcome
Ui fails to load, then resyncronizes UI.
The only error message is in the javascript console `(TypeError) : Cannot read properties of undefined (reading 'a')` and `Uncaught (in promise) Error: Client is resynchronizing`

### Expected outcome

The ui loads as expected, or at least an error is thrown.

### Minimal reproducible example

```java
package com.example.application.views.main;

import com.vaadin.flow.component.dialog.Dialog;
import com.vaadin.flow.component.grid.Grid;
import com.vaadin.flow.component.orderedlayout.HorizontalLayout;
import com.vaadin.flow.router.PageTitle;
import com.vaadin.flow.router.Route;

import java.util.ArrayList;

@PageTitle("Main")
@Route(value = "")
public class MainView extends HorizontalLayout {
    private static class Item {
        public String name;

        public Item(String s) {
            name = s;
        }
    }

    public MainView() {
        setSizeFull();

        var items = new ArrayList<Item>();

        for (int i = 0; i < 100; i++) {
            items.add(new Item("Item " + i));
        }

        var grid = new Grid<Item>();
        grid.setSizeFull();
        grid.setItems(query -> items.stream().skip(query.getOffset()).limit(query.getLimit()));
        grid.addColumn(item -> item.name);

        grid.addAttachListener(event -> {
            // this makes everything fail to load
            event.getUI().push();
        });

        var dialog = new Dialog();
        dialog.add(grid);
        dialog.setSizeFull();
        dialog.open();
    }
}

```

### Steps to reproduce

1. Download the Vaadin 24.0.4 starter project
2. Add `@Push(PushMode.MANUAL)` to the Application.java
3. Replace the MainView.java with the provided example
4. Notice the Dialog is not loading properly
5. Comment out `event.getUI().push();` and see how everything load correctly.
6. Additionally, notice that `event.getUI().accessSynchronously(() -> event.getUI().push());` fails aswell, while `event.getUI().access(() -> event.getUI().push());` seems to delay the push but work

### Environment

Vaadin version(s): 24.0.4
OS: Windows 11 x64

**Comment by mrgreywater** [comment]
I'll be honest, I have no idea what a UIDL request is, and I don't really see any exceptions when to call the push method, as long as it's in a vaadin thread. This also worked fine in Vaadin 23.

The usecase here is that I'm loading a large (expensive) grid. Many rows, many columns. If I'd just call setItems in the constructor/when creating the grid, it would take quite a while and the browser would be somwhat unresponsive before it even shows anything. So I'm splitting the task. First I create the grid without the items. I make sure the grid (with column headers, but no items) is shown by calling ui.push() in the attach-listener. Then (also in the attach listener) I call grid.setItems() and another ui.push() to start populating the grid.

Basically it just makes it feel more responsive for the user.

**Comment by mcollovati** [comment]
Below a quick, simplified and not accurate explanation of what happens.
When the page is loaded or when the user interacts with some component, an HTTP request is sent to perform the actions on the server side; UI changes are serialized in a specific format (UIDL) in the HTTP response and applied by the client. Every request and response carries server and client identifiers to ensure that changes are processed in the correct order, otherwise a full resynchronization is triggered.

In your case, there's a there's a request from the browser to initialize the view, and starts to build a message with server sync id A.
However, before the sync id is incremented and the response is sent, the `push` call in the attach listener is called (as it is a change part of the initiator request) and a new message is created with the same server sync id (A). The sync id is then incremented by the  `push` invocation (B = A + 1) and again by the completion of the initiator request (C= B + 1 = A + 2), before writing the response.
Thus, the client receives two messages with the same server id A and discards the second one, as it is considered already seen.
On the next interaction, the client expects a message with server id B, but it instead receives C, and after a while waiting for B it requests a full resynchronization.

![image](https://user-images.githubusercontent.com/4648894/233559064-a5dbc2e7-79fe-4438-8354-b7a109e1e033.png)


This situation is definitely something that the framework should handle or prevent (e.g. by throwing an exception), and, as you said, the documentation should probably be improved

While we investigate further on the issue, I suggest you to push the heavy changes from a separated thread.
This way the UI can be initialized early without freezing and changes from heavy computation can be correctly pushed to update the client.

```java
grid.addAttachListener(event -> {
    UI ui = event.getUI();
    CompletableFuture.runAsync(() -> {
        ui.access(() -> {
            Notification.show("Start loading data");
        });
        List<Item> data = heavyLoading();
        ui.access(() -> {
            grid.setItems(data);
            ui.push();
        });
        ui.access(() -> {
            Notification.show("Data loaded");
        });
    });
});
```

**Comment by mrgreywater** [comment]
Thanks for your explanation and workaround suggestion.

The async thread is not really needed in my case, as the data is already provided with an AbstractDataProvider. The freezing the users would experience is more in the javascript side, as it's loading a bunch of elements in the dom synchronously. Splitting them up makes them load one after another.

My workaround for now is as follows:
```java
        grid.addAttachListener(event -> {
            event.getUI().access(() -> {
                event.getUI().push();
                grid.setItems(dataProviderWithFilter);
                grid.recalculateColumnWidths();
                event.getUI().push();
            });
        });
```

**Comment by TatuLund** [comment]
There is actually built-in feature in DataCommunicator to use background threads for updates

https://github.com/vaadin/flow/blob/main/flow-data/src/main/java/com/vaadin/flow/data/provider/DataCommunicator.java#L359

This allows 1-liner implementation, i.e. you do not need to create futures etc. I would assume it also enables your target of showing Grid eagerly without data, before data is being fetched.

**Comment by TatuLund** [comment]
ExecutorService executorService = Executors.newSingleThreadExecutor();

**Comment by mrgreywater** [comment]
Not really in my case, my DataProvider needs the current session, so I need to assign the CurrentInstances to the new thread. Also the way you're doing it swallows all exceptions of the dataprovider (see my comment above).
